File: macos/Sources/GlyphstrokeCore/Geometry.swift
import Foundation

/// Точка росчерка в экранной системе координат: X вправо, Y вниз.
///
/// Своя структура, а не CGPoint: ядро не зависит от графических рамок
/// системы, и его тесты гоняются где угодно — в том числе там, где
/// CoreGraphics нет вовсе.
public struct Point: Equatable, Sendable {
    public var x: Double
    public var y: Double

    public init(_ x: Double, _ y: Double) {
        self.x = x
        self.y = y
    }

    public func distance(to other: Point) -> Double {
        hypot(other.x - x, other.y - y)
    }
}

/// Геометрия росчерка: всё, что нужно алгоритму $1.
///
/// Перенесено из версии для Linux (`glyphstroke/recognizer.py`) один в один —
/// числа и пороги обязаны совпадать, иначе один и тот же жест на двух
/// машинах распознавался бы по-разному, а файлы жестов у нас общие.
public enum Geometry {
    /// Сколько точек оставляем после выравнивания по длине пути.
    public static let numPoints = 64
    /// Сторона квадрата, в который вписывается росчерк перед сравнением.
    public static let squareSize = 250.0
    public static let halfDiagonal = 0.5 * hypot(squareSize, squareSize)
    /// Золотое сечение — для поиска лучшего угла.
    public static let phi = 0.5 * (-1.0 + 5.0.squareRoot())
    /// Росчерк вытянутее, чем 1:4, сжимаем равномерно: иначе прямая линия
    /// растянулась бы в квадрат и потеряла форму.
    public static let thinRatio = 0.25

    public static func pathLength(_ points: [Point]) -> Double {
        guard points.count > 1 else { return 0 }
        var total = 0.0
        for i in 1..<points.count {
            total += points[i - 1].distance(to: points[i])
        }
        return total
    }

    public static func dedupe(_ points: [Point], eps: Double = 1e-9) -> [Point] {
        var out: [Point] = []
        for p in points where out.isEmpty || out[out.count - 1].distance(to: p) > eps {
            out.append(p)
        }
        return out
    }

    /// Разложить росчерк на `n` равноудалённых по длине пути точек.
    public static func resample(_ points: [Point], n: Int = numPoints) -> [Point] {
        var pts = dedupe(points)
        guard pts.count >= 2 else {
            return Array(repeating: pts.first ?? Point(0, 0), count: n)
        }
        let interval = pathLength(pts) / Double(n - 1)
        guard interval > 0 else { return Array(repeating: pts[0], count: n) }

        var out: [Point] = [pts[0]]
        var accumulated = 0.0
        var i = 1
        while i < pts.count {
            let d = pts[i - 1].distance(to: pts[i])
            if accumulated + d >= interval {
                let t = (interval - accumulated) / d
                let next = Point(
                    pts[i - 1].x + t * (pts[i].x - pts[i - 1].x),
                    pts[i - 1].y + t * (pts[i].y - pts[i - 1].y)
                )
                out.append(next)
                // Точка вставляется в исходный список: следующий отрезок
                // отмеряется уже от неё, иначе шаг «уплывает».
                pts.insert(next, at: i)
                accumulated = 0
            } else {
                accumulated += d
            }
            i += 1
        }
        // Хвост добираем из-за накопленной ошибки вещественных чисел.
        while out.count < n {
            out.append(pts[pts.count - 1])
        }
        return Array(out.prefix(n))
    }

    public static func centroid(_ points: [Point]) -> Point {
        guard !points.isEmpty else { return Point(0, 0) }
        let count = Double(points.count)
        let sum = points.reduce(Point(0, 0)) { Point($0.x + $1.x, $0.y + $1.y) }
        return Point(sum.x / count, sum.y / count)
    }

    /// Левый верхний угол и размеры охватывающего прямоугольника.
    public static func boundingBox(_ points: [Point]) -> (x: Double, y: Double, width: Double, height: Double) {
        guard !points.isEmpty else { return (0, 0, 0, 0) }
        let xs = points.map(\.x)
        let ys = points.map(\.y)
        let minX = xs.min()!, maxX = xs.max()!
        let minY = ys.min()!, maxY = ys.max()!
        return (minX, minY, maxX - minX, maxY - minY)
    }

    public static func rotateBy(_ points: [Point], radians: Double) -> [Point] {
        let c = centroid(points)
        let cosA = cos(radians), sinA = sin(radians)
        return points.map { p in
            Point(
                (p.x - c.x) * cosA - (p.y - c.y) * sinA + c.x,
                (p.x - c.x) * sinA + (p.y - c.y) * cosA + c.y
            )
        }
    }

    public static func scaleToSquare(_ points: [Point], size: Double = squareSize) -> [Point] {
        let box = boundingBox(points)
        if box.width <= 0 && box.height <= 0 { return points }
        let longest = max(box.width, box.height)
        let thin = longest > 0 ? min(box.width, box.height) / longest < thinRatio : true
        if thin {
            // Вытянутый росчерк (прямая, узкая дуга) — только равномерное сжатие.
            let k = size / longest
            return points.map { Point($0.x * k, $0.y * k) }
        }
        return points.map { Point($0.x * size / box.width, $0.y * size / box.height) }
    }

    public static func translateToOrigin(_ points: [Point]) -> [Point] {
        let c = centroid(points)
        return points.map { Point($0.x - c.x, $0.y - c.y) }
    }

    /// Привести росчерк к виду, пригодному для сравнения с шаблоном.
    public static func normalize(_ points: [Point], n: Int = numPoints) -> [Point] {
        translateToOrigin(scaleToSquare(resample(points, n: n)))
    }

    public static func pathDistance(_ a: [Point], _ b: [Point]) -> Double {
        let count = min(a.count, b.count)
        guard count > 0 else { return .infinity }
        var total = 0.0
        for i in 0..<count {
            total += a[i].distance(to: b[i])
        }
        return total / Double(a.count)
    }

    public static func distanceAtAngle(_ points: [Point], template: [Point], angle: Double) -> Double {
        pathDistance(rotateBy(points, radians: angle), template)
    }

    /// Поиск золотым сечением по углу в пределах ±`tolerance`.
    ///
    /// Допуск маленький (по умолчанию 20°): он гасит дрожание руки, но не
    /// позволяет спутать жест с его же поворотом на 90°.
    public static func distanceAtBestAngle(
        _ points: [Point],
        template: [Point],
        tolerance: Double,
        precision: Double = 2.0 * .pi / 180.0
    ) -> Double {
        let straight = distanceAtAngle(points, template: template, angle: 0)
        guard tolerance > 0 else { return straight }

        var lo = -tolerance, hi = tolerance
        var x1 = phi * lo + (1.0 - phi) * hi
        var f1 = distanceAtAngle(points, template: template, angle: x1)
        var x2 = (1.0 - phi) * lo + phi * hi
        var f2 = distanceAtAngle(points, template: template, angle: x2)

        while abs(hi - lo) > precision {
            if f1 < f2 {
                hi = x2
                x2 = x1
                f2 = f1
                x1 = phi * lo + (1.0 - phi) * hi
                f1 = distanceAtAngle(points, template: template, angle: x1)
            } else {
                lo = x1
                x1 = x2
                f1 = f2
                x2 = (1.0 - phi) * lo + phi * hi
                f2 = distanceAtAngle(points, template: template, angle: x2)
            }
        }
        // Угол 0 проверяем отдельно: поиск в него не попадает и слегка портит
        // оценку даже при точном совпадении росчерка с шаблоном.
        return min(straight, min(f1, f2))
    }
}

File: windows/src/Glyphstroke.Core/Geometry.cs
namespace Glyphstroke.Core;

/// <summary>Точка росчерка: X вправо, Y вниз — как на экране.</summary>
/// <remarks>
/// Своя структура, а не System.Drawing.PointF: ядро не зависит от графических
/// библиотек, и его проверки гоняются где угодно.
/// </remarks>
public readonly record struct Point(double X, double Y)
{
    public double DistanceTo(Point other) => Math.Sqrt(
        (other.X - X) * (other.X - X) + (other.Y - Y) * (other.Y - Y));
}

/// <summary>
/// Геометрия росчерка: всё, что нужно алгоритму $1 Unistroke.
/// </summary>
/// <remarks>
/// Перенесено из версии для Linux (<c>glyphstroke/recognizer.py</c>) число в число.
/// Это не педантизм: файлы жестов у трёх версий общие, и разъехавшиеся пороги
/// означали бы, что один и тот же жест на двух машинах ведёт себя по-разному,
/// причём молча.
/// </remarks>
public static class Geometry
{
    /// <summary>Сколько точек остаётся после выравнивания по длине пути.</summary>
    public const int NumPoints = 64;

    /// <summary>Сторона квадрата, в который вписывается росчерк.</summary>
    public const double SquareSize = 250.0;

    public static readonly double HalfDiagonal = 0.5 * Math.Sqrt(SquareSize * SquareSize * 2);

    /// <summary>Золотое сечение — для поиска лучшего угла.</summary>
    public static readonly double Phi = 0.5 * (-1.0 + Math.Sqrt(5.0));

    /// <summary>
    /// Росчерк вытянутее, чем 1:4, сжимаем равномерно: иначе прямая линия
    /// растянулась бы в квадрат и потеряла форму.
    /// </summary>
    public const double ThinRatio = 0.25;

    public static double PathLength(IReadOnlyList<Point> points)
    {
        double total = 0;
        for (int i = 1; i < points.Count; i++)
        {
            total += points[i - 1].DistanceTo(points[i]);
        }
        return total;
    }

    public static List<Point> Dedupe(IReadOnlyList<Point> points, double eps = 1e-9)
    {
        var result = new List<Point>(points.Count);
        foreach (var point in points)
        {
            if (result.Count == 0 || result[^1].DistanceTo(point) > eps)
            {
                result.Add(point);
            }
        }
        return result;
    }

    /// <summary>Разложить росчерк на <paramref name="n"/> равноудалённых точек.</summary>
    public static List<Point> Resample(IReadOnlyList<Point> points, int n = NumPoints)
    {
        var pts = Dedupe(points);
        if (pts.Count < 2)
        {
            var single = pts.Count == 1 ? pts[0] : new Point(0, 0);
            return Enumerable.Repeat(single, n).ToList();
        }

        double interval = PathLength(pts) / (n - 1);
        if (interval <= 0)
        {
            return Enumerable.Repeat(pts[0], n).ToList();
        }

        var output = new List<Point> { pts[0] };
        double accumulated = 0;
        int i = 1;
        while (i < pts.Count)
        {
            double d = pts[i - 1].DistanceTo(pts[i]);
            if (accumulated + d >= interval)
            {
                double t = (interval - accumulated) / d;
                var next = new Point(
                    pts[i - 1].X + t * (pts[i].X - pts[i - 1].X),
                    pts[i - 1].Y + t * (pts[i].Y - pts[i - 1].Y));
                output.Add(next);
                // Точка вставляется в исходный список: следующий отрезок
                // отмеряется уже от неё, иначе шаг «уплывает».
                pts.Insert(i, next);
                accumulated = 0;
            }
            else
            {
                accumulated += d;
            }
            i++;
        }

        // Хвост добираем из-за накопленной ошибки вещественных чисел.
        while (output.Count < n)
        {
            output.Add(pts[^1]);
        }
        return output.Take(n).ToList();
    }

    public static Point Centroid(IReadOnlyList<Point> points)
    {
        if (points.Count == 0)
        {
            return new Point(0, 0);
        }
        double x = 0, y = 0;
        foreach (var point in points)
        {
            x += point.X;
            y += point.Y;
        }
        return new Point(x / points.Count, y / points.Count);
    }

    public static (double X, double Y, double Width, double Height) BoundingBox(IReadOnlyList<Point> points)
    {
        if (points.Count == 0)
        {
            return (0, 0, 0, 0);
        }
        double minX = points.Min(p => p.X), maxX = points.Max(p => p.X);
        double minY = points.Min(p => p.Y), maxY = points.Max(p => p.Y);
        return (minX, minY, maxX - minX, maxY - minY);
    }

    public static List<Point> RotateBy(IReadOnlyList<Point> points, double radians)
    {
        var c = Centroid(points);
        double cos = Math.Cos(radians), sin = Math.Sin(radians);
        return points.Select(p => new Point(
            (p.X - c.X) * cos - (p.Y - c.Y) * sin + c.X,
            (p.X - c.X) * sin + (p.Y - c.Y) * cos + c.Y)).ToList();
    }

    public static List<Point> ScaleToSquare(IReadOnlyList<Point> points, double size = SquareSize)
    {
        var box = BoundingBox(points);
        if (box.Width <= 0 && box.Height <= 0)
        {
            return points.ToList();
        }

        double longest = Math.Max(box.Width, box.Height);
        bool thin = longest > 0 ? Math.Min(box.Width, box.Height) / longest < ThinRatio : true;
        if (thin)
        {
            double k = size / longest;
            return points.Select(p => new Point(p.X * k, p.Y * k)).ToList();
        }
        return points.Select(p => new Point(p.X * size / box.Width, p.Y * size / box.Height)).ToList();
    }

    public static List<Point> TranslateToOrigin(IReadOnlyList<Point> points)
    {
        var c = Centroid(points);
        return points.Select(p => new Point(p.X - c.X, p.Y - c.Y)).ToList();
    }

    /// <summary>Привести росчерк к виду, пригодному для сравнения с шаблоном.</summary>
    public static List<Point> Normalize(IReadOnlyList<Point> points, int n = NumPoints) =>
        TranslateToOrigin(ScaleToSquare(Resample(points, n)));

    public static double PathDistance(IReadOnlyList<Point> a, IReadOnlyList<Point> b)
    {
        int count = Math.Min(a.Count, b.Count);
        if (count == 0)
        {
            return double.PositiveInfinity;
        }
        double total = 0;
        for (int i = 0; i < count; i++)
        {
            total += a[i].DistanceTo(b[i]);
        }
        return total / a.Count;
    }

    public static double DistanceAtAngle(IReadOnlyList<Point> points, IReadOnlyList<Point> template, double angle) =>
        PathDistance(RotateBy(points, angle), template);

    /// <summary>
    /// Поиск золотым сечением по углу в пределах ±<paramref name="tolerance"/>.
    /// </summary>
    /// <remarks>
    /// Допуск маленький (по умолчанию 20°): он гасит дрожание руки, но не даёт
    /// спутать жест с его же поворотом на 90°.
    /// </remarks>
    public static double DistanceAtBestAngle(IReadOnlyList<Point> points, IReadOnlyList<Point> template,
        double tolerance, double precision = 2.0 * Math.PI / 180.0)
    {
        double straight = DistanceAtAngle(points, template, 0);
        if (tolerance <= 0)
        {
            return straight;
        }

        double lo = -tolerance, hi = tolerance;
        double x1 = Phi * lo + (1.0 - Phi) * hi;
        double f1 = DistanceAtAngle(points, template, x1);
        double x2 = (1.0 - Phi) * lo + Phi * hi;
        double f2 = DistanceAtAngle(points, template, x2);

        while (Math.Abs(hi - lo) > precision)
        {
            if (f1 < f2)
            {
                hi = x2;
                x2 = x1;
                f2 = f1;
                x1 = Phi * lo + (1.0 - Phi) * hi;
                f1 = DistanceAtAngle(points, template, x1);
            }
            else
            {
                lo = x1;
                x1 = x2;
                f1 = f2;
                x2 = (1.0 - Phi) * lo + Phi * hi;
                f2 = DistanceAtAngle(points, template, x2);
            }
        }
        // Угол 0 проверяем отдельно: поиск в него не попадает и слегка портит
        // оценку даже при точном совпадении росчерка с шаблоном.
        return Math.Min(straight, Math.Min(f1, f2));
    }
}
